lib/trie/trie.go

```diff
@@ -80,52 +82,50 @@ func (t *Trie) handleTrackedDeltas(success bool, pendingDeletedMerkleValues map[
 	}
 }
 
-func (t *Trie) prepLeafForMutation(currentLeaf *Node,
+func (t *Trie) prepForMutation(currentNode *Node,
 	copySettings node.CopySettings,
-	pendingDeletedMerkleValues map[string]struct{}) (newLeaf *Node) {
-	if currentLeaf.Generation == t.generation {
-		// no need to deep copy and update generation
-		// of current leaf.
-		newLeaf = currentLeaf
-	} else {
-		newLeaf = updateGeneration(currentLeaf, t.generation, pendingDeletedMerkleValues, copySettings)
+	pendingDeletedMerkleValues map[string]struct{}) (
+	newNode *Node, err error) {
+	isRoot := currentNode == t.root
+	err = registerDeletedMerkleValue(currentNode, isRoot,
+		pendingDeletedMerkleValues)
+	if err != nil {
+		return nil, fmt.Errorf("registering deleted node: %w", err)
 	}
-	newLeaf.SetDirty()
-	return newLeaf
-}
 
-func (t *Trie) prepBranchForMutation(currentBranch *Node,
-	copySettings node.CopySettings,
-	pendingDeletedMerkleValues map[string]struct{}) (newBranch *Node) {
-	if currentBranch.Generation == t.generation {
+	if currentNode.Generation == t.generation {
 		// no need to deep copy and update generation
-		// of current branch.
-		newBranch = currentBranch
+		// of current node.
+		newNode = currentNode
 	} else {
-		newBranch = updateGeneration(currentBranch, t.generation, pendingDeletedMerkleValues, copySettings)
+		newNode = currentNode.Copy(copySettings)
+		newNode.Generation = t.generation
 	}
-	newBranch.SetDirty()
-	return newBranch
+	newNode.SetDirty()
+	return newNode, nil
 }
 
-// updateGeneration is called when the currentNode is from
-// an older trie generation (snapshot) so we deep copy the
-// node and update the generation on the newer copy.
-func updateGeneration(currentNode *Node, trieGeneration uint64,
-	deletedMerkleValues map[string]struct{}, copySettings node.CopySettings) (
-	newNode *Node) {
-	newNode = currentNode.Copy(copySettings)
-	newNode.Generation = trieGeneration
+func registerDeletedMerkleValue(node *Node, isRoot bool,
+	pendingDeletedMerkleValues map[string]struct{}) (err error) {
+	err = ensureMerkleValueIsCalculated(node, isRoot)
+	if err != nil {
+		return fmt.Errorf("ensuring Merkle value is calculated: %w", err)
+	}
+
+	if len(node.MerkleValue) < 32 {
+		// Merkle values which are less than 32 bytes are inlined
+		// in the parent branch and are not stored on disk, so there
+		// is no need to track their deletion for the online pruning.
+		return nil
+	}
 
-	// The hash of the node from a previous snapshotted trie
-	// is usually already computed.
-	deletedMerkleValue := currentNode.MerkleValue
-	if len(deletedMerkleValue) > 0 {
-		deletedMerkleValueString := string(deletedMerkleValue)
-		deletedMerkleValues[deletedMerkleValueString] = struct{}{}
+	if !node.Dirty {
+		// Only register deleted nodes that were not previously modified
+		// since the last trie snapshot.
+		pendingDeletedMerkleValues[string(node.MerkleValue)] = struct{}{}
```

Comment:
why don't we just add this key directly to `t.deletedMerkleValues`? I don't see the reason we are passing around this map of `deletedMerkleValues` which will eventually be used to update `t.deletedMerkleValues` in `t.handleTrackedDeltas`.

Reply:
This was implemented in #2876 in order to not have tracked trie deltas in a bad state in case of error during a single trie operation (method call). On top of that, this is especially important for error handling with the trie lazy loading since we IO from disk and each call returns an error (which may fail at any 'depth' in the trie).